akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala

```diff
@@ -632,10 +633,15 @@ Host: example.com
       val clientSettings = ConnectionPoolSettings(system).withConnectionSettings(ClientConnectionSettings(system).withSocketOptions(SO.ReceiveBufferSize(serverToClientNetworkBufferSize) :: Nil))
 
       val serverConnectionContext = ExampleHttpContexts.exampleServerContext
-      // Disable hostname verification as ExampleHttpContexts.exampleClientContext sets hostname as akka.example.org
-      val sslConfigSettings = SSLConfigSettings().withLoose(SSLLooseConfig().withDisableHostnameVerification(true))
-      val sslConfig = AkkaSSLConfig().withSettings(sslConfigSettings)
-      val clientConnectionContext = ConnectionContext.https(ExampleHttpContexts.exampleClientContext.sslContext, Some(sslConfig))
+
+      // Create SSLEngine without hostname verification as ExampleHttpContexts.exampleServerContext has a certificate
+      // that is valid for the 'akka.example.org' but we're connecting to 'localhost' - so we need to disable hostname verification.
+      val insecureSslEngineFactory = () => {
+        val engine = ExampleHttpContexts.exampleClientSSLContext.createSSLEngine()
+        engine.setUseClientMode(true)
```

Comment:
We could, though that would obscure an error when the user tries to use a client-mode SSLEngine for a server-side connection or vice-versa.

Also, because SSLEngine is mutable, this could lead to confusing problems when people are reusing an SSLEngine. Since the SSLEngine is connection-specific anyway, though, that wouldn't work in the first place so that's probably fine.

I think I agree it's helpful to set this on 'our side'.

Reply:
Then we should probably also do `engine.getSSLParameters.setEndpointIdentificationAlgorithm("https")` for them, and make using a custom ClientTransport the only way to disable hostname verification.

Reply:
I think we should do as much as possible. E.g. I found that it matters when you call `engine.setUseClientMode`. If you do it to late, former parameter settings might be silently discarded (I noticed while testing client-side ALPN for HTTP/2).

Reply:
> it matters when you call engine.setUseClientMode. If you do it to late, former parameter settings might be silently discarded (I noticed while testing client-side ALPN for HTTP/2)

Ouch - doesn't that mean we shouldn't set it as late as `sslTlsServerStage`, since that might parameters set by users?

Reply:
... and of course this is also different between jdk8 and jdk11. after out-of-band discussion: to be on the safe side, let's update the client mode in `sslTlsServerStage`, but only when needed.

Reply:
Unfortunately 'only when needed' is not possible on JDK11: `getUseClientMode` does return a value when the mode isn't set yet, but if you then try to use the engine, it throws because the mode wasn't set yet :/.

Reply:
Ok, this was thought to be a usability improvement, but if it's not possible, what can you do?